window functions docs (#14739)
apache/druid: docs/querying/sql-window-functions.md

  ~ distributed with this work for additional information
  ~ regarding copyright ownership.  The ASF licenses this file
  ~ to you under the Apache License, Version 2.0 (the
  ~ License); you may not use this file except in compliance
  ~ with the License.  You may obtain a copy of the License at
  ~
  ~   http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing,
  ~ software distributed under the License is distributed on an
  ~ AS IS BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->

> Apache Druid supports two query languages: [Druid SQL](sql.md) and [native queries](querying.md).
> This document describes the SQL language.
>
> Window functions are an [experimental](../development/experimental.md) feature. Development and testing are still at early stage. Feel free to try window functions and provide your feedback.
> There are known issues where ORDER BY only works on ascending order and certain options may cause errors.

> > Set the context parameter `windowsAreForClosers: true` to use window functions.

Reviewer: Instead of having this in a note, could this go into a Prerequisites section?

Author: It's not exactly a pre-requisite. You do it per-query. And it's like a feature flag that will go away once the feature is AF

Reviewer: ```suggestion
> To use window functions, submit the context parameter `windowsAreForClosers: true` with your query.
```

Reviewer: Maybe also link to context parameters in case someone doesn't know how to set it?

Window functions in Apache Druid produce values based upon the relationship of one row within a window of rows to the other rows within the same window. A window is a group of related rows within a result set. For example, rows with the same value for a specific dimension.

> The following example returns the rank of each row in ascending order based upon its `delta` value as related to other rows with the same `channel` value.

Reviewer: ```suggestion
The following example groups results with the same `channel` value into windows. For each window, the query returns the rank of each row in ascending order based upon its `delta` value.
```

| event_time | channel | change | rank_value |
| -- | -- | -- | -- |
| 2016-06-27T00:00:00.000Z	| #kk.wikipedia	| 1	| 1 |
| 2016-06-27T00:00:00.000Z	| #kk.wikipedia	| 1	| 1 |
| 2016-06-27T00:00:00.000Z	| #kk.wikipedia	| 7	| 3 |
| 2016-06-27T00:00:00.000Z	| #kk.wikipedia	| 56 | 4 |
| 2016-06-27T00:00:00.000Z	| #kk.wikipedia	| 56 | 4 |
| 2016-06-27T00:00:00.000Z	| #kk.wikipedia	| 63 | 6 |
| 2016-06-27T00:00:00.000Z	| #kk.wikipedia	| 91 | 7 |  
| 2016-06-27T00:00:00.000Z	| #kk.wikipedia	| 2440 | 8 |
| 2016-06-27T00:00:00.000Z	| #kk.wikipedia	| 2703 | 9 |
| 2016-06-27T00:00:00.000Z	| #kk.wikipedia	| 6900 |10 |
| 2016-06-27T00:00:00.000Z	| #lt.wikipedia	| 1	| 1 |
| 2016-06-27T00:00:00.000Z	| #lt.wikipedia	| 2	| 2 |
| 2016-06-27T00:00:00.000Z	| #lt.wikipedia	| 13 | 3 |
| 2016-06-27T00:00:00.000Z	| #lt.wikipedia	| 28 | 4 |
| 2016-06-27T00:00:00.000Z	| #lt.wikipedia	| 53 | 5 |
| 2016-06-27T00:00:00.000Z	| #lt.wikipedia	| 56 | 6 |
| 2016-06-27T00:00:00.000Z	| #lt.wikipedia	| 59 | 7 |
| 2016-06-27T00:00:00.000Z	| #lt.wikipedia	| 391 | 8 |
| 2016-06-27T00:00:00.000Z	| #lt.wikipedia	| 894	| 9 |
| 2016-06-27T00:00:00.000Z	| #lt.wikipedia	| 4358 | 10 |

</details>

> Window functions are similar to [aggregation funtctions](./aggregations.md).  

Reviewer: ```suggestion
Window functions are similar to [aggregation functions](./aggregations.md).  
```

> You can use the OVER clause treat other Druid aggretation functions as window functions. For example, the sum of a value for rows within a window.

Reviewer: ```suggestion
You can use the OVER clause to treat other Druid aggregation functions as window functions. For example, the sum of a value for rows within a window.
```

When working with window functions, consider the following:
- Window functions only work on GROUP BY queries.
- Window functions support aliasing.

## Define a window with the OVER clause

The OVER clause defines the query windows for window functions as follows:
- PARTITION BY indicates the dimension that defines the rows within the window
- ORDER BY specifies the order of the rows within the windows.
Sometimes windows are called partitions, but don't confuse them with the segment partitioning feature of Druid.

> The PARTITIONED BY clause defines the dimenson to use for the window. For example, the following OVER clause sets the window dimension to `channel` and orders the results by the absolute value of `delta` ascending:

Reviewer: Duplicate?
>The PARTITIONED BY clause defines the dimenson to use for the window.

```sql
...

> rank() OVER (PARTITION BY channel ORDER BY ABS(delta) ASC)

Reviewer: Add an explanation on why you would use this query instead of the one above, even though they are functionally the same. Because the previous one is a GROUP BY but this one can be used in other queries such as...?

Author: I don't understand. This is not a query, just an example of an OVER clause. If you wrote out the whole query, it would need to be a GROUP BY. The only difference here is that the window definition is inline (PARTITION.... ) instead of being aliased (w) in the previous example

Reviewer: Oh I see that, never mind my comment.

> rank() OVER (PARTITION BY channel ORDER BY ABS(delta) ASC)

Reviewer: nit: consistency on casing between this query and the preceding query. See `rank/RANK` and `abs/ABS`

...
```

## Window function reference

> |Function|Notes|Default|

Reviewer: Do you need a default column if all the defaults here are None?

Author: Do we need it for the UI. Not sure.

|--------|-----|-------|
| `ROW_NUMBER()`| Returns the number of the row within the window| None |
|`RANK()`| Returns the rank for a row within a window | None | 

> |`DENSE_RANK()`| Returns the ranks for a row within a window without gaps. For example, if two rows tie for rank of 1, the subsequent row is ranked 2. | None |

Reviewer: ```suggestion
|`DENSE_RANK()`| Returns the rank for a row within a window without gaps. For example, if two rows tie for rank of 1, the subsequent row is ranked 2. | None |
```
Does Druid do any kind of sorting for which tied row gets returned first?

|`PERCENT_RANK()`| Returns the rank of the row calculated as a percentage according to the formula: `(rank - 1) / (total partition rows - 1)` | None |

> |`CUME_DIST()`| Returns the cumulative distribution of the current row within the window calculated as `number of partition rows preceding or peers with current row` / `total partition rows` | None |

Reviewer: >preceding or peers with current row

Maybe "at the same rank or higher as the current row"?

> |`CUME_DIST()`| Returns the cumulative distribution of the current row within the window calculated as `number of partition rows preceding or peers with current row` / `total partition rows` | None |

Reviewer: Since a lot of the descriptions use the term "window" instead of "partition" the formulas should say window as well

Author: gret catch

> |`NTILE(tiles)`| Divides the results as evenly as possible into the number of tiles, also called buckets, and returns a value from 1 to the value of `tiles`  |None |

Reviewer: I'm not sure I understand what this function does even after rereading it a few times. Are the results divided per each window? If I call `NTILES(5)`, it returns a result from 1 to 5, but how do I know what will be returned?

> |`LAG(expr, offset, default)`| Returns the value evaluated at the row that precedes the current row by the offset number within the window. If there is no such row, returns the default. | None |

Reviewer: ```suggestion
|`LAG(expr, offset, default)`| Returns the value evaluated at the row that precedes the current row by the offset number within the window; if there is no such row, returns the given default value | None |
```

> |`LEAD(expr, offset, default)`| Returns the value evaluated at the row that followsthe current row by the offset number within the window; if there is no such row, returns the default | None |

Reviewer: ```suggestion
|`LEAD(expr, offset, default)`| Returns the value evaluated at the row that follows the current row by the offset number within the window; if there is no such row, returns the given default value | None |
```

|`FIRST_VALUE(expr)`| Returns the value for the expression for the first row within the window| None |

> |`LAST_VALUE(expr)`| Returns the value for the expression for the first row within the window | None |

Reviewer: ```suggestion
|`LAST_VALUE(expr)`| Returns the value for the expression for the last row within the window | None |
```

## Example

The following example illustrates all of the built-in window functions to compare the number of characters changed per event for a channel in the Wikipedia data set.

```sql
SELECT FLOOR(__time TO DAY) AS event_time,
    channel,
    ABS(delta) AS change,
    ROW_NUMBER() OVER w AS row_no,
    RANK() OVER w AS rank_no,
    DENSE_RANK() OVER w AS dense_rank_no,
    PERCENT_RANK() OVER w AS pct_rank,

>     CUME_DIST() OVER w AS cumlative_dist,

Reviewer: ```suggestion
    CUME_DIST() OVER w AS cumulative_dist,
```

    NTILE(4) OVER w AS ntile_val,
    LAG(ABS(delta), 1, 0) OVER w AS lag_val,
    LEAD(ABS(delta), 1, 0) OVER w AS lead_val,
    FIRST_VALUE(ABS(delta)) OVER w AS first_val,
    LAST_VALUE(ABS(delta)) OVER w AS last_val
FROM wikipedia
WHERE channel IN ('#kk.wikipedia', '#lt.wikipedia')
GROUP BY channel, ABS(delta), FLOOR(__time TO DAY) 
WINDOW w AS (PARTITION BY channel ORDER BY ABS(delta) ASC)
```

<details>
<summary> View results </summary>

> |event_time|channel|change|row_no|rank_no|dense_rank_no|pct_rank|cumlative_dist|ntile_val|lag_val|lead_val|first_val|last_val|

Reviewer: ```suggestion
|event_time|channel|change|row_no|rank_no|dense_rank_no|pct_rank|cumulative_dist|ntile_val|lag_val|lead_val|first_val|last_val|
```

> This document describes the SQL language.
>
> Window functions are an [experimental](../development/experimental.md) feature. Development and testing are still at early stage. Feel free to try window functions and provide your feedback.
> There are known issues where ORDER BY only works on ascending order and certain options may cause errors.
> Set the context parameter `windowsAreForClosers: true` to use window functions.

Window functions in Apache Druid produce values based upon the relationship of one row within a window of rows to the other rows within the same window. A window is a group of related rows within a result set. For example, rows with the same value for a specific dimension.

The following example groups results with the same `channel` value into windows. For each window, the query returns the rank of each row in ascending order based upon its `delta` value.

```sql
SELECT FLOOR(__time TO DAY) AS event_time,
    channel,
    ABS(delta) AS change,
    RANK() OVER w AS rank_value
FROM wikipedia
WHERE channel in ('`#kk.wikipedia`', '#lt.wikipedia')
AND '2016-06-28' > FLOOR(__time TO DAY) > '2016-06-26'
GROUP BY channel, ABS(delta), __time

> WINDOW w AS (PARTITION BY channel ORDER BY ABS(delta) ASC)

Reviewer 2: My understanding is only one of the ordering (either ASC or DESC) works. Can you double check which one works and call this out as a limitation?

|`2016-06-27T04:35:00.000Z`|`#kk.wikipedia`|2440|2496|
|`2016-06-27T06:15:00.000Z`|`#kk.wikipedia`|91|2587|
|`2016-06-27T07:32:00.000Z`|`#kk.wikipedia`|1|2588|
|`2016-06-27T09:00:00.000Z`|`#kk.wikipedia`|2703|5291|
|`2016-06-27T09:24:00.000Z`|`#kk.wikipedia`|1|5292|
|`2016-06-27T11:00:00.000Z`|`#kk.wikipedia`|63|5355|
|`2016-06-27T11:05:00.000Z`|`#kk.wikipedia`|7|5362|
|`2016-06-27T11:32:00.000Z`|`#kk.wikipedia`|56|5418|
|`2016-06-27T15:21:00.000Z`|`#kk.wikipedia`|6900|12318|
|`2016-06-27T06:17:00.000Z`|`#lt.wikipedia`|2|2|
|`2016-06-27T07:55:00.000Z`|`#lt.wikipedia`|13|15|
|`2016-06-27T09:05:00.000Z`|`#lt.wikipedia`|894|909|
|`2016-06-27T09:12:00.000Z`|`#lt.wikipedia`|391|1300|
|`2016-06-27T09:23:00.000Z`|`#lt.wikipedia`|56|1356|
|`2016-06-27T10:59:00.000Z`|`#lt.wikipedia`|1|1357|
|`2016-06-27T11:49:00.000Z`|`#lt.wikipedia`|59|1416|
|`2016-06-27T12:41:00.000Z`|`#lt.wikipedia`|53|1469|
|`2016-06-27T12:58:00.000Z`|`#lt.wikipedia`|28|1497|
|`2016-06-27T19:03:00.000Z`|`#lt.wikipedia`|4358|5855|

</details>

## Known issues

The following are known issues with window functions:

- Descending order, DESC, in the ORDER BY clause in the window definition causes a Null Pointer Exception.

> - Druid throws an exception if you do not include the offset and default for LEAD() and LAG() functions.

Reviewer 3: What would the alternative be?  That is, if the offset is not defined, what should be returned?  The default can be `null`, so that makes sense, but I'm unclear what a meaning default value for `offset` in `LEAD`/`LAG` would be?

Reviewer 4: this default offset stuff was fixed recently

|`PERCENT_RANK()`| Returns the rank of the row calculated as a percentage according to the formula: `(rank - 1) / (total window rows - 1)` |
|`CUME_DIST()`| Returns the cumulative distribution of the current row within the window calculated as `number of window rows at the same rank or higher than current row` / `total window rows` |
|`NTILE(tiles)`| Divides the rows within a window as evenly as possible into the number of tiles, also called buckets, and returns the value of the tile that the row falls into  |None |
|`LAG(expr, offset, default)`| Returns the value evaluated at the row that precedes the current row by the offset number within the window; if there is no such row, returns the given default value |
|`LEAD(expr, offset, default)`| Returns the value evaluated at the row that follows the current row by the offset number within the window; if there is no such row, returns the given default value |
|`FIRST_VALUE(expr)`| Returns the value for the expression for the first row within the window|
|`LAST_VALUE(expr)`| Returns the value for the expression for the last row within the window |

> ## Example

Author: ```suggestion
## Examples
```

---
id: sql-window-functions
title: Window functions
---

<!--
  ~ Licensed to the Apache Software Foundation (ASF) under one
  ~ or more contributor license agreements.  See the NOTICE file
  ~ distributed with this work for additional information
  ~ regarding copyright ownership.  The ASF licenses this file
  ~ to you under the Apache License, Version 2.0 (the
  ~ License); you may not use this file except in compliance
  ~ with the License.  You may obtain a copy of the License at
  ~
  ~   http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing,
  ~ software distributed under the License is distributed on an
  ~ AS IS BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->

:::info

Apache Druid supports two query languages: [Druid SQL](sql.md) and [native queries](querying.md).
This document describes the SQL language.

> Window functions are an [experimental](../development/experimental.md) feature. Development and testing are still at early stage. Feel free to try window functions and provide your feedback.

Reviewer 5: ```suggestion
Window functions are an [experimental](../development/experimental.md) feature. Development and testing are still at early stage. Feel free to try window functions and provide your feedback. Windows functions are not supported by multi-stage-query engine and thus cannot be used in SQL-based Ingestion as of now. 
```

> Window functions are an [experimental](../development/experimental.md) feature. Development and testing are still at early stage. Feel free to try window functions and provide your feedback.

Author: ```suggestion
Window functions are an [experimental](../development/experimental.md) feature.
Development and testing are still at early stage. Feel free to try window functions and provide your feedback.
Window functions are an [experimental](../development/experimental.md) feature.
Feel free to try window functions and provide your feedback.
Windows functions are not currently supported by multi-stage-query engine so you cannot use them in SQL-based ingestion. 
```

> There are known issues where ORDER BY only works on ascending order and certain options may cause errors.

Reviewer 6: This has been fixed recently. Now both the orders work fine. The known issue mentioned [here](https://docs.imply.io/latest/druid/querying/sql-window-functions/#known-issues) has been resolved

> There are known issues where ORDER BY only works on ascending order and certain options may cause errors.

Author: ```suggestion
```

| `2016-06-27T00:00:00.000Z`| `#kk.wikipedia`| 2703 | 9 |
| `2016-06-27T00:00:00.000Z`| `#kk.wikipedia`| 6900 |10 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 1 | 1 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 2 | 2 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 13 | 3 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 28 | 4 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 53 | 5 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 56 | 6 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 59 | 7 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 391 | 8 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 894 | 9 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 4358 | 10 |

</details>

Window functions are similar to [aggregation functions](./aggregations.md).  

You can use the OVER clause to treat other Druid aggregation functions as window functions. For example, the sum of a value for rows within a window.

Window functions support aliasing.

## Define a window with the OVER clause

The OVER clause defines the query windows for window functions as follows:
- PARTITION BY indicates the dimension that defines the rows within the window

> - ORDER BY specifies the order of the rows within the windows. Currently only ascending order, ASC, works.

Author: ```suggestion
- ORDER BY specifies the order of the rows within the windows.
```

|------|---------|---------|-------------|
|`2016-06-27T04:20:00.000Z`|`#kk.wikipedia`|56|56|
|`2016-06-27T04:35:00.000Z`|`#kk.wikipedia`|2440|2496|
|`2016-06-27T06:15:00.000Z`|`#kk.wikipedia`|91|2587|
|`2016-06-27T07:32:00.000Z`|`#kk.wikipedia`|1|2588|
|`2016-06-27T09:00:00.000Z`|`#kk.wikipedia`|2703|5291|
|`2016-06-27T09:24:00.000Z`|`#kk.wikipedia`|1|5292|
|`2016-06-27T11:00:00.000Z`|`#kk.wikipedia`|63|5355|
|`2016-06-27T11:05:00.000Z`|`#kk.wikipedia`|7|5362|
|`2016-06-27T11:32:00.000Z`|`#kk.wikipedia`|56|5418|
|`2016-06-27T15:21:00.000Z`|`#kk.wikipedia`|6900|12318|
|`2016-06-27T06:17:00.000Z`|`#lt.wikipedia`|2|2|
|`2016-06-27T07:55:00.000Z`|`#lt.wikipedia`|13|15|
|`2016-06-27T09:05:00.000Z`|`#lt.wikipedia`|894|909|
|`2016-06-27T09:12:00.000Z`|`#lt.wikipedia`|391|1300|
|`2016-06-27T09:23:00.000Z`|`#lt.wikipedia`|56|1356|
|`2016-06-27T10:59:00.000Z`|`#lt.wikipedia`|1|1357|
|`2016-06-27T11:49:00.000Z`|`#lt.wikipedia`|59|1416|
|`2016-06-27T12:41:00.000Z`|`#lt.wikipedia`|53|1469|
|`2016-06-27T12:58:00.000Z`|`#lt.wikipedia`|28|1497|
|`2016-06-27T19:03:00.000Z`|`#lt.wikipedia`|4358|5855|

</details>

## Known issues

The following are known issues with window functions:

> - Descending order, DESC, in the ORDER BY clause in the window definition causes a Null Pointer Exception.

Reviewer 4: no; this is not there anymore - however the following is known:

* aggregates which have `ORDER BY` specified are processed in the window: `ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW` 
     * note: other database seem to use the default of  `RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW`  in this case (not sure if it is specified or not in the standard - have to dig it out)
     * note: in case the order column is unique there is no difference between `RANGE` / `ROWS` 
* windows with `RANGE` specifications are handled as `ROWS`
* `LEAD/LAG` ignores the default value
* `LAST_VALUE` returns the last value of the window even in the presence of an `order by` 

here is an example query which could illustrate these (not sure if needed)
```
  SELECT
    FLOOR(__time TO DAY) t,
    cnt,
    FLOOR(m1/3),
    COUNT(1) OVER (ORDER BY FLOOR(m1/3) ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW), --ok
    COUNT(1) OVER (ORDER BY FLOOR(m1/3) RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW), -- bad
    LAST_VALUE(cnt) OVER (ORDER BY __time), -- last value of the window
    LAG(cnt) OVER (),  -- ok
    LAG(cnt,1,999) OVER () -- default value ignored
  FROM foo

```

---
id: sql-window-functions
title: Window functions
---

<!--
  ~ Licensed to the Apache Software Foundation (ASF) under one
  ~ or more contributor license agreements.  See the NOTICE file
  ~ distributed with this work for additional information
  ~ regarding copyright ownership.  The ASF licenses this file
  ~ to you under the Apache License, Version 2.0 (the
  ~ License); you may not use this file except in compliance
  ~ with the License.  You may obtain a copy of the License at
  ~
  ~   http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing,
  ~ software distributed under the License is distributed on an
  ~ AS IS BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->

:::info

Apache Druid supports two query languages: [Druid SQL](sql.md) and [native queries](querying.md).

> This document describes the SQL language.

Reviewer 4: I'm not sure if we need these lines ; I believe this file will open from `SQL` sidebard stuff; so I feel they are  pretty redundant

Author: This is in all the SQL reference docs. This admonition is here for all the SQL docs, so it should be in this one too. We can consider removing as an effort to remove from all SQL

Window functions are an [experimental](../development/experimental.md) feature. Development and testing are still at early stage. Feel free to try window functions and provide your feedback.

There are known issues where ORDER BY only works on ascending order and certain options may cause errors.

> Set the context parameter `windowsAreForClosers: true` to use window functions.

Reviewer 4: we should rename this to something else :)

| `2016-06-27T00:00:00.000Z`| `#kk.wikipedia`| 63 | 6 |
| `2016-06-27T00:00:00.000Z`| `#kk.wikipedia`| 91 | 7 |  
| `2016-06-27T00:00:00.000Z`| `#kk.wikipedia`| 2440 | 8 |
| `2016-06-27T00:00:00.000Z`| `#kk.wikipedia`| 2703 | 9 |
| `2016-06-27T00:00:00.000Z`| `#kk.wikipedia`| 6900 |10 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 1 | 1 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 2 | 2 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 13 | 3 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 28 | 4 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 53 | 5 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 56 | 6 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 59 | 7 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 391 | 8 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 894 | 9 |
| `2016-06-27T00:00:00.000Z`| `#lt.wikipedia`| 4358 | 10 |

</details>

Window functions are similar to [aggregation functions](./aggregations.md).  

You can use the OVER clause to treat other Druid aggregation functions as window functions. For example, the sum of a value for rows within a window.

When working with window functions, consider the following:

> - Window functions only work on GROUP BY queries.

Reviewer 4: I don't think we need this anymore - it should work

Reviewer 6: I tried a couple of queries like 
```
SELECT	
		RANK() OVER (PARTITION BY m1 ORDER BY m1 DESC) 
			AS ranking,
		m1
FROM foo 
```
Does not give any results but if I add a `GROUP BY m1` it does give results. As for this I would like to include GROUP BY in the docs as the window functions are still experimental and once we are confident that the cases run without the group by we can update it in the next release

- Window functions support aliasing.

## Define a window with the OVER clause

The OVER clause defines the query windows for window functions as follows:
- PARTITION BY indicates the dimension that defines the rows within the window
- ORDER BY specifies the order of the rows within the windows. Currently only ascending order, ASC, works.

:::note

> Sometimes windows are called partitions. However, the partitioning for window functions are a shuffle or a partition of the result set that happens at query time and is not to be confused with Druid's segment partitioning feature which partitions data at ingest time.

Author: ```suggestion
Sometimes windows are called partitions. However, the partitioning for window functions are a shuffle (partition) of the result set that happens at query time and is not to be confused with Druid's segment partitioning feature which partitions data at ingest time.
```

:::

The following OVER clause example sets the window dimension to `channel` and orders the results by the absolute value of `delta` ascending:

```sql
...
RANK() OVER (PARTITION BY channel ORDER BY ABS(delta) ASC)
...
```

## Window function reference

|Function|Notes|
|--------|-----|
| `ROW_NUMBER()`| Returns the number of the row within the window|
|`RANK()`| Returns the rank for a row within a window | 
|`DENSE_RANK()`| Returns the rank for a row within a window without gaps. For example, if two rows tie for rank of 1, the subsequent row is ranked 2. |
|`PERCENT_RANK()`| Returns the rank of the row calculated as a percentage according to the formula: `(rank - 1) / (total window rows - 1)` |
|`CUME_DIST()`| Returns the cumulative distribution of the current row within the window calculated as `number of window rows at the same rank or higher than current row` / `total window rows` |
|`NTILE(tiles)`| Divides the rows within a window as evenly as possible into the number of tiles, also called buckets, and returns the value of the tile that the row falls into  |None |

> |`LAG(expr, offset, default)`| Returns the value evaluated at the row that precedes the current row by the offset number within the window; if there is no such row, returns the given default value |

Reviewer 4: I don't think we have `LAG(expr, offset, default)` ; to my best knowledge we have [this](https://github.com/apache/druid/blob/d25caaefa4282a6b61663e8184d3c0b4092f6f35/sql/src/main/java/org/apache/druid/sql/calcite/rel/Windowing.java#L91)

for which we could write:
* `LAG(expr, [offset])`
* `offset` defaults to `1` if omitted
* remove the part about default value  

similar commment notes for `LEAD`

|`LEAD(expr, offset, default)`| Returns the value evaluated at the row that follows the current row by the offset number within the window; if there is no such row, returns the given default value |
|`FIRST_VALUE(expr)`| Returns the value for the expression for the first row within the window|
|`LAST_VALUE(expr)`| Returns the value for the expression for the last row within the window |

## Examples

The following example illustrates all of the built-in window functions to compare the number of characters changed per event for a channel in the Wikipedia data set.

```sql
SELECT FLOOR(__time TO DAY) AS event_time,
    channel,
    ABS(delta) AS change,
    ROW_NUMBER() OVER w AS row_no,
    RANK() OVER w AS rank_no,
    DENSE_RANK() OVER w AS dense_rank_no,
    PERCENT_RANK() OVER w AS pct_rank,
    CUME_DIST() OVER w AS cumulative_dist,
    NTILE(4) OVER w AS ntile_val,

>     LAG(ABS(delta), 1, 0) OVER w AS lag_val,

Reviewer 4: I don't think this `default` work - I see `null` in the output; and not `0`
I assume this query can be run so I think we accept it ; but doesn't live up to working accordingly....

Author: @kgyrtkirk , don't know if this is a bug in the function or if there is not an intention to be a default value (I was basing my docs off pg)

Reviewer 4: its a bug in the `LEAD/LAG` implementation ; it doesn't take the `default` into consideration; however `Calcite` accepts it - we could either reject these ; or implement it - either way I've opened [this ticket](https://github.com/apache/druid/issues/15270) for it

|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|53|5|5|5|0.4444444444444444|0.5|2|28|56|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|56|6|6|6|0.5555555555555556|0.6|2|53|59|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|59|7|7|7|0.6666666666666666|0.7|3|56|391|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|391|8|8|8|0.7777777777777778|0.8|3|59|894|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|894|9|9|9|0.8888888888888888|0.9|4|391|4358|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|4358|10|10|10|1|1|4|894|null|1|4358|

</details>

The following example demonstrates applying a window to the SUM() function to calculate the cumulative changes to a channel over time:

```sql
SELECT
    FLOOR(__time TO MINUTE) as "time",
    channel,
    ABS(delta) AS changes,
    sum(ABS(delta)) OVER (PARTITION BY channel ORDER BY FLOOR(__time TO HOUR) ASC) AS cum_changes
FROM wikipedia
WHERE channel IN ('#kk.wikipedia', '#lt.wikipedia')
GROUP BY channel, __time, delta
```

<details>
<summary> View results </summary>

|`time`|`channel`|`changes`|`cum_changes`|
|------|---------|---------|-------------|
|`2016-06-27T04:20:00.000Z`|`#kk.wikipedia`|56|56|

> |`2016-06-27T04:35:00.000Z`|`#kk.wikipedia`|2440|2496|

Reviewer 4: I might have made a mistake I can't see right now - but I'm pretty confident that this should be `2496` for both of these rows

I think we have:
* `OVER(... ORDER BY FLOOR( __time TO HOUR)` => should be ordered by the floored value
* running aggregate of `SUM()` should be calculated for up-to-all current values of orderby value...
* these 2 rows should floor to 4am but they have different running-sum values 

 
I was playing with pg in [this sqlfiddle](http://sqlfiddle.com/#!17/8aac4f/3)

Author: @kgyrtkirk 

The query floors to `MINUTE` not hour in line 177:
`FLOOR(__time TO MINUTE)`

For consistency I changed line 180 to floor to `MINUTE` too -- it keeps the same results as I have here. 

I wonder if there is a bug with Druid in the case that you pointed out. Where it looks like it honored the `MINUTE` and not the `HOUR`?

Reviewer 4: for the resultset `line 180` mattered more :) 

Thank you for changing it now the resultset seem to be more accurate! :dagger:

---
id: sql-window-functions
title: Window functions
---

<!--
  ~ Licensed to the Apache Software Foundation (ASF) under one
  ~ or more contributor license agreements.  See the NOTICE file
  ~ distributed with this work for additional information
  ~ regarding copyright ownership.  The ASF licenses this file
  ~ to you under the Apache License, Version 2.0 (the
  ~ License); you may not use this file except in compliance
  ~ with the License.  You may obtain a copy of the License at
  ~
  ~   http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing,
  ~ software distributed under the License is distributed on an
  ~ AS IS BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->

:::info

Apache Druid supports two query languages: [Druid SQL](sql.md) and [native queries](querying.md).
This document describes the SQL language.

> Window functions are an [experimental](../development/experimental.md) feature.

Reviewer 6: There is a repetition of the statement here. I dont think it is expected

|`PERCENT_RANK()`| Returns the rank of the row calculated as a percentage according to the formula: `(rank - 1) / (total window rows - 1)` |
|`CUME_DIST()`| Returns the cumulative distribution of the current row within the window calculated as `number of window rows at the same rank or higher than current row` / `total window rows` |
|`NTILE(tiles)`| Divides the rows within a window as evenly as possible into the number of tiles, also called buckets, and returns the value of the tile that the row falls into | None |
|`LAG(expr[, offset])`| Returns the value evaluated at the row that precedes the current row by the offset number within the window. `offset` defaults to 1 if not provided |
|`LEAD(expr[, offset])`| Returns the value evaluated at the row that follows the current row by the offset number within the window; if there is no such row, returns the given default value. `offset` defaults to 1 if not provided |
|`FIRST_VALUE(expr)`| Returns the value for the expression for the first row within the window |
|`LAST_VALUE(expr)`| Returns the value for the expression for the last row within the window |

## Examples

> The following example illustrates all of the built-in window functions to compare the number of characters changed per event for a channel in the Wikipedia data set.

Reviewer 6: Probably sometime later when these are not experimental we could run an example with a use case and show how an empty over() works followed by introducing partition by and order by similar to how Drill does in https://drill.apache.org/docs/sql-window-functions-introduction/. Not now, but some food for thought for future

|`2016-06-27T00:00:00.000Z`|`#kk.wikipedia`|2440|6|6|6|0.7142857142857143|0.75|3|91|2703|1|6900|
|`2016-06-27T00:00:00.000Z`|`#kk.wikipedia`|2703|7|7|7|0.8571428571428571|0.875|4|2440|6900|1|6900|
|`2016-06-27T00:00:00.000Z`|`#kk.wikipedia`|6900|8|8|8|1|1|4|2703|null|1|6900|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|1|1|1|1|0|0.1|1|null|2|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|2|2|2|2|0.1111111111111111|0.2|1|1|13|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|13|3|3|3|0.2222222222222222|0.3|1|2|28|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|28|4|4|4|0.3333333333333333|0.4|2|13|53|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|53|5|5|5|0.4444444444444444|0.5|2|28|56|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|56|6|6|6|0.5555555555555556|0.6|2|53|59|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|59|7|7|7|0.6666666666666666|0.7|3|56|391|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|391|8|8|8|0.7777777777777778|0.8|3|59|894|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|894|9|9|9|0.8888888888888888|0.9|4|391|4358|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|4358|10|10|10|1|1|4|894|null|1|4358|

</details>

> The following example demonstrates applying a window to the SUM() function to calculate the cumulative changes to a channel over time:

Reviewer 6: Might rephrase as applying the sum() function on the values in the window

    ABS(delta) AS change,
    RANK() OVER w AS rank_value
FROM wikipedia

> WHERE channel in ('`#kk.wikipedia`', '#lt.wikipedia')

Reviewer: ```suggestion
WHERE channel in ('#kk.wikipedia', '#lt.wikipedia')
```

The OVER clause defines the query windows for window functions as follows:
- PARTITION BY indicates the dimension that defines the rows within the window
- ORDER BY specifies the order of the rows within the windows. Currently only ascending order, ASC, works.

:::note

> Sometimes windows are called partitions. However, the partitioning for window functions are a shuffle/partition of the result set that happens at query time and is not to be confused with Druid's segment partitioning feature which partitions data at ingest time.

Reviewer: Recommend removing slashes for clarity: https://developers.google.com/style/slashes#slashes-with-alternatives

    FIRST_VALUE(ABS(delta)) OVER w AS first_val,
    LAST_VALUE(ABS(delta)) OVER w AS last_val
FROM wikipedia

> WHERE channel IN ('`#kk.wikipedia`', '#lt.wikipedia')

Reviewer: ```suggestion
WHERE channel IN ('#kk.wikipedia', '#lt.wikipedia')
```

|`2016-06-27T00:00:00.000Z`|`#kk.wikipedia`|2440|6|6|6|0.7142857142857143|0.75|3|91|2703|1|6900|
|`2016-06-27T00:00:00.000Z`|`#kk.wikipedia`|2703|7|7|7|0.8571428571428571|0.875|4|2440|6900|1|6900|
|`2016-06-27T00:00:00.000Z`|`#kk.wikipedia`|6900|8|8|8|1|1|4|2703|null|1|6900|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|1|1|1|1|0|0.1|1|null|2|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|2|2|2|2|0.1111111111111111|0.2|1|1|13|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|13|3|3|3|0.2222222222222222|0.3|1|2|28|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|28|4|4|4|0.3333333333333333|0.4|2|13|53|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|53|5|5|5|0.4444444444444444|0.5|2|28|56|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|56|6|6|6|0.5555555555555556|0.6|2|53|59|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|59|7|7|7|0.6666666666666666|0.7|3|56|391|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|391|8|8|8|0.7777777777777778|0.8|3|59|894|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|894|9|9|9|0.8888888888888888|0.9|4|391|4358|1|4358|
|`2016-06-27T00:00:00.000Z`| `#lt.wikipedia`|4358|10|10|10|1|1|4|894|null|1|4358|

</details>

> The following example demonstrates applying a window to the SUM() function to calcualte the cumulative changes to a channel over time:

Reviewer: ```suggestion
The following example demonstrates applying a window to the SUM() function to calculate the cumulative changes to a channel over time:
```

> ```

Reviewer: `````suggestion
```sql
`````
